### PR TITLE
WIP: Add initial syscall/js.copyBytesToJS support

### PR DESCRIPTION
**Pull request by justinclift**
It's working in initial testing, but could likely be improved somewhat. :smile:

### CODE REVIEW

**Comment by deadprogram** [comment]
Any chance of an automated test that exercises this?

**Comment by justinclift** [comment]
Good idea.  Probably best to wait a few days, to see if it works ok with the further WebGL stuff.  There's every chance it'll need tweaking/adjusting (etc). :wink:

File: targets/wasm_exec.js [hunk 1]
```diff
@@ -313,6 +313,26 @@
 					//"syscall/js.valueInstanceOf": (sp) => {
 					//	mem().setUint8(sp + 24, loadValue(sp + 8) instanceof loadValue(sp + 16));
 					//},
+
+					// copyBytesToJS(dst ref, src []byte) (int, bool)
+					// Originally copied from upstream Go project, then modified:
+					//   https://github.com/golang/go/blob/3f995c3f3b43033013013e6c7ccc93a9b1411ca9/misc/wasm/wasm_exec.js#L404-L416
+					// param4 and param5 (below) are likely "length" variables of some sort
+					"syscall/js.copyBytesToJS": (ret_addr, dest_addr, source_addr, param4, param5) => {
```
Comment (aykevl) [inline targets/wasm_exec.js]: The calling convention changes have changed this function conceptually to the following:

```go
func copyBytesToJS(returnValue *struct{int; bool}, dst *ref, srcPtr *byte, srcLen, srcCap uintptr)
```

The following two rules apply to it:

  * large return values are passed as a pointer-to-struct and prepended to the parameter list
  * `int64`/`uint64` (the `ref` type) are passed as a pointer-to-value
  * slices are split into 3 parts: the pointer to the underlying array, the length, and the capacity.

Comment (justinclift) [reply targets/wasm_exec.js]: Thanks.  I'll update things accordingly over the next few days. :smile:

Comment (aykevl) [reply targets/wasm_exec.js]: Any chance you can take a look at this PR? The next release is getting close.

Comment (justinclift) [reply targets/wasm_exec.js]: Yep, I can make time this weekend. :smile:

**Comment by deadprogram** [comment]
Is this PR still valid? if so, any chance of it being able to bring it back to life?

**Comment by justinclift** [comment]
No idea, haven't touched Go at all in months.  Might as well close it, and if there turns out to be need of it in future it can be re-opened or re-submitted.

**Comment by justinclift** [comment]
@deadprogram Looking back at this now, it turns out this PR is still needed after all. :wink:

**Comment by justinclift** [comment]
Wasn't able to re-open this PR for some unknown reason (GitHub technical issue?), so recreated it in #933.

**Pull request closed without merging**